Add privacy/security doc for delay rendering. (#730)
MicrosoftEdge/MSEdgeExplainers: DelayedClipboard/tag-security-privacy-delayed-clipboard-rendering.md

### Questions from https://www.w3.org/TR/security-privacy-questionnaire/

## 2. Questions to Consider

### 2.1. What information might this feature expose to Web sites or other parties, and for what purposes is that exposure necessary?

> For standard formats, this feature doesn't expose any information about the source and destination apps during copy-paste operation. However, for web custom formats, since it's restricted to a specific set of apps, during paste it could expose the source app where the user copied the data from. When a format is delay rendered, the system clipboard requests the data for that format during paste. The source app must provide the data for the delay rendered format when it is requested by the system clipboard, else, the paste would fail in the destination app.

Reviewer: ```suggestion
For standard formats, this feature doesn't expose any information about the source and destination apps during copy-paste operation. However, for web custom formats, since some apps may use ecosystem-specific web custom formats, during paste it could expose the source app where the user copied the data from. When a format is delay rendered, the system clipboard requests the data for that format during paste. The source app must provide the data for the delay rendered format when it is requested by the system clipboard, else, the paste would fail in the destination app.
```

Author: Done.

For more discussion, please see https://github.com/w3c/editing/issues/439.

To use delay rendering via the async clipboard API, all the existing restrictions in async clipboard APIs related to secure context, permissions etc apply.

### 2.2. Do features in your specification expose the minimum amount of information necessary to enable their intended uses?

Yes.

> ### 2.3. How do the features in your specification deal with personal information, personally-identifiable information (PII), or information derived from them?

Reviewer: Is this true for delay rendering of web custom formats?

Author: I wasn't sure if a trusted app (where the user is performing the paste) knowing where the user copied the data from would fall under PII. I'm not sure how to describe that part so I mentioned it as No.

Author: Done.

### 2.14. How do the features in this specification work in the context of a browser’s Private Browsing or Incognito mode?

It works the same way in incognito mode as the existing async clipboard APIs.

### 2.15. Does this specification have both "Security Considerations" and "Privacy Considerations" sections?

Yes. https://github.com/snianu/MSEdgeExplainers/blob/main/DelayedClipboard/DelayedClipboardRenderingExplainer.md#privacy-and-security-considerations.

### 2.16. Do features in your specification enable origins to downgrade default security protections?

No.

### 2.17. How does your feature handle non-"fully active" documents?

No interaction with documents regardless of its state. However, the document needs to have focus to trigger the existing async clipboard APIs.

### 2.18. What should this questionnaire have asked?

N/A

## 3. Threat Models

### 3.1. Passive Network Attackers

No threat.

### 3.2. Active Network Attackers

The API is only available from a secure context.

### 3.3. Same-Origin Policy Violations

It doesn't leak data across origins.

### 3.4. Third-Party Tracking

No interaction with third-party pages.

> ### 3.5. Legitimate Misuse

Reviewer: The mitigation of adding a limit doesn't fully eliminate misuse of the API, it just prevents a broad net from being cast by attackers. Perhaps we should make that more explicit.

Author: Done.

> We don't think there is any risk of legitimate misuse of this API. The privacy concern with web custom formats can be mitigated by restricting the web custom formats to a very small number so the authors can't cast a wide net to track all the supported apps. UAs could also choose to trigger the callback after a fixed/variable amount of time regardless of whether the formats were requested by the system clipboard or not. That way the authors can't determine whether the data is being requested as a result of user paste operation.

Reviewer: For "UAs could also choose to trigger the callback after a fixed/variable amount of time regardless of whether the formats were requested by the system clipboard or not. That way the authors can't determine whether the data is being requested as a result of user paste operation.", it might help to move that into a separate paragraph and elaborate more on how it ambiguates the paste results but also that it has tradeoffs in terms of performance.

Author: Done.

### 2.16. Do features in your specification enable origins to downgrade default security protections?

No.

### 2.17. How does your feature handle non-"fully active" documents?

No interaction with documents regardless of its state. However, the document needs to have focus to trigger the existing async clipboard APIs.

### 2.18. What should this questionnaire have asked?

N/A

## 3. Threat Models

### 3.1. Passive Network Attackers

No threat.

### 3.2. Active Network Attackers

The API is only available from a secure context.

### 3.3. Same-Origin Policy Violations

It doesn't leak data across origins.

### 3.4. Third-Party Tracking

No interaction with third-party pages.

### 3.5. Legitimate Misuse

The privacy concern with web custom formats can be partially mitigated by restricting the web custom formats to a very small number so the authors can't cast a wide net to track all the supported apps. It doesn't address the core privacy issue where a target app is able to know where the user has copied the data from if there is a limited number of apps that support a web custom format.

> UAs could also choose to trigger the callback after a fixed/variable amount of time regardless of whether the formats were requested by the system clipboard or not. That way the authors can't determine whether the data is being requested as a result of user paste operation. This prevents web authors from truly delay rendering the web custom formats and it affects performance as typically web custom formats have high fidelity content.

Reviewer: ```suggestion
UAs could also choose to trigger the callback after a fixed/variable amount of time regardless of whether the formats were requested by the system clipboard or not. That way the authors can't determine whether the data is being requested as a result of user paste operation. This, however, is a tradeoff as it may prevent sites from getting the full benefits of delay rendering for web custom formats, as that would require the payload to be generated only on-demand.
```

Author: Done.
